tokenawarebandit_test.go

```diff
@@ -0,0 +1,367 @@
+		{hostId: "11", connectAddress: net.IPv4(10, 0, 0, 12), tokens: []string{"60"}, dataCenter: "remote2"},
+	}
+	policy.AddHosts(hosts[:])
+
+	policy.SetPartitioner("OrderedPartitioner")
+
+	policy.getKeyspaceMetadata = func(keyspaceName string) (*KeyspaceMetadata, error) {
+		if keyspaceName != keyspace {
+			return nil, fmt.Errorf("unknown keyspace: %s", keyspaceName)
+		}
+		return &KeyspaceMetadata{
+			Name:          keyspace,
+			StrategyClass: "NetworkTopologyStrategy",
+			StrategyOptions: map[string]interface{}{
+				"class":   "NetworkTopologyStrategy",
+				"local":   2,
+				"remote1": 2,
+				"remote2": 2,
+			},
+		}, nil
+	}
+	replicasUpdateChan := make(chan struct{})
+	policy.afterReplicasUpdated = func(keyspaceName string) {
+		if keyspaceName == keyspace {
+			close(replicasUpdateChan)
+		}
+	}
+	policy.KeyspaceChanged(KeyspaceUpdateEvent{Keyspace: keyspace})
+	<-replicasUpdateChan
+
+	// The NetworkTopologyStrategy above should generate the following replicas.
+	// It's handy to have as reference here.
+	assertDeepEqual(t, "replicas", tokenRingReplicas{
+		{orderedToken("05"), []*HostInfo{hosts[0], hosts[1], hosts[2], hosts[3], hosts[4], hosts[5]}},
+		{orderedToken("10"), []*HostInfo{hosts[1], hosts[2], hosts[3], hosts[4], hosts[5], hosts[6]}},
+		{orderedToken("15"), []*HostInfo{hosts[2], hosts[3], hosts[4], hosts[5], hosts[6], hosts[7]}},
+		{orderedToken("20"), []*HostInfo{hosts[3], hosts[4], hosts[5], hosts[6], hosts[7], hosts[8]}},
+		{orderedToken("25"), []*HostInfo{hosts[4], hosts[5], hosts[6], hosts[7], hosts[8], hosts[9]}},
+		{orderedToken("30"), []*HostInfo{hosts[5], hosts[6], hosts[7], hosts[8], hosts[9], hosts[10]}},
+		{orderedToken("35"), []*HostInfo{hosts[6], hosts[7], hosts[8], hosts[9], hosts[10], hosts[11]}},
+		{orderedToken("40"), []*HostInfo{hosts[7], hosts[8], hosts[9], hosts[10], hosts[11], hosts[0]}},
+		{orderedToken("45"), []*HostInfo{hosts[8], hosts[9], hosts[10], hosts[11], hosts[0], hosts[1]}},
+		{orderedToken("50"), []*HostInfo{hosts[9], hosts[10], hosts[11], hosts[0], hosts[1], hosts[2]}},
+		{orderedToken("55"), []*HostInfo{hosts[10], hosts[11], hosts[0], hosts[1], hosts[2], hosts[3]}},
+		{orderedToken("60"), []*HostInfo{hosts[11], hosts[0], hosts[1], hosts[2], hosts[3], hosts[4]}},
+	}, policy.keyspaces[keyspace].replicas)
+	if len(policy.keyspaces) != 1 {
+		t.Fatalf("expected only single keyspace, got %v", policy.keyspaces)
+	}
+
+	// now the token ring is configured
+	query.RoutingKey([]byte("18"))
+	iter = policy.Pick(query)
+
+	selectedHostList := iterToSlice(iter)
+	if len(selectedHostList) != 12 {
+		t.Fatalf("expected 12 hosts, got %v", selectedHostList)
+	}
+	sort.Strings(selectedHostList[2:6])  // order of non-local replicas does not matter
+	sort.Strings(selectedHostList[6:12]) // order of other hosts does not matter
+	expected := []string{"4", "7", "3", "5", "6", "8", "0", "1", "10", "11", "2", "9"}
```

Comment:
This confused me for a moment because `sort.Strings("10", "2") == []string{"10", "2"}`. Do you think we could make all the host IDs two-digit, i.e. `"02"` instead of `"2"`?

Reply:
Good idea! Updated.